daft/dashboard.py

```diff
@@ -49,41 +48,63 @@ def _should_run() -> bool:
     return True
 
 
-def _broadcast_query_plan(
+class NativeLogBroadcaster:
+    def __init__(self):
+        self.original_stdout = sys.stdout
+        self.buffer = ""
+
+    def write(self, text: str):
+        self.original_stdout.write(text)
+        self.buffer += text
+
+    def flush(self):
+        self.original_stdout.flush()
+
+    def close(self):
+        self.original_stdout.close()
+
+
+def broadcast_query_information(
     mermaid_plan: str,
     plan_time_start: datetime,
     plan_time_end: datetime,
 ):
-    dashboard = dashboard_module()
-
-    # try launching the dashboard
-    # if dashboard is already launched, this will do nothing
-    dashboard.launch(detach=True, noop_if_initialized=True)
-
-    headers = {
-        "Content-Type": "application/json",
-    }
-    data = json.dumps(
-        {
-            "id": str(uuid.uuid4()),
-            "mermaid_plan": mermaid_plan,
-            "plan_time_start": str(plan_time_start),
-            "plan_time_end": str(plan_time_end),
-        }
-    ).encode("utf-8")
-    req = request.Request(dashboard.DAFT_DASHBOARD_QUERIES_URL, headers=headers, data=data)
+    dashboard = _dashboard_module()
+    dashboard.broadcast_query_information(
+        mermaid_plan=mermaid_plan,
+        plan_time_start=plan_time_start,
+        plan_time_end=plan_time_end,
+        logs=_get_logs(),
+    )
 
-    try:
-        request.urlopen(req, timeout=1)
-    except URLError as e:
-        warnings.warn(f"Failed to broadcast metrics over {dashboard.DAFT_DASHBOARD_QUERIES_URL}: {e}")
+
+log_capturer = None
+
+
+def _get_logs():
+    if not log_capturer:
+        return ""
+
+    return log_capturer.buffer
 
 
 try:
-    dashboard = dashboard_module()
+    dashboard = _dashboard_module()
     launch = dashboard.launch
     shutdown = dashboard.shutdown
 
+    runner = get_context().get_or_create_runner()
+
+    if runner.name == "ray":
+        pass
+    elif runner.name == "py":
+        pass
+    elif runner.name == "native":
+        log_capturer = NativeLogBroadcaster()
+
+    if log_capturer:
+        sys.stdout = log_capturer
```

Comment:
This is where the magic of capturing `STDOUT` occurs. All `STDOUT` is redirected to my `NativeLogBroadcaster` class, which prints the printout to console, but also captures it in a buffer.